### ISSUE TITLE
nixos/zsh prefers completions shipped with Zsh over completions shipped with installed packages

### ISSUE THREAD

**Issue by eigengrau** [NixOS/nixpkgs#197502]
### Describe the bug
The NixOS Zsh module currently sets the fpath in a way in which completions shipped with Zsh are preferred over completions included in an installed package.

### Steps To Reproduce

1. Deploy a NixOS system.

```nix
programs.zsh = {
  enable = true;
  enableCompletion = true;
};
```

2. Type `curl -<TAB>`
3. Observe that no completions are shown, because the completion shipped with Zsh does not handle options (compare with `curl <TAB>`)

### Expected behavior
Completions shipped with packages installed into profiles should be preferred over completions shipped with Zsh.

### Notify maintainers
No maintainers are set, so I’ll just ping @Mic92 and @nrdxp.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.74, NixOS, 22.05 (Quokka), 22.05.20221022.3933d8b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.0`
 - channels(root): `""`
 - channels(seb): `""`
 - nixpkgs: `/etc/nixpkgs/channels/stable`
```

**Comment by nrdxp** [comment]
@eigengrau, curl might be a bad example of this, since just adding `environment.systemPackages = [ curl ]` does not include the completions for curl in your system derivation at all. You have to also include `curl.out` (since the default output is `bin`) in this case. This is a fair bit annoying and non-obvious, but completions do work if you do this.

**Comment by eigengrau** [comment]
@nrdxp I don’t quite get it, yet. It looks like you are correct that the reproduction steps are insufficient because of the mulit-output situation. However, if you extend the reproducer accordingly, you should notice that Zsh will still not use completions provided by the curl package (and the code makes it clear why this is the case). Could you confirm, e.g., via `nixos-shell --flake .\#vm`?

```nix
  {
    outputs = { nixpkgs, ... }:
      let
        lib = nixpkgs.lib;
        module = { pkgs, ... }: {
          environment.systemPackages = [ pkgs.curl.out ];
          programs.zsh = {
            enable = true;
            enableCompletion = true;
          };
        };
      in
      {
        nixosConfigurations.vm = lib.makeOverridable lib.nixosSystem {
          modules = [ module ];
          system = "x86_64-linux";
        };
      };
```

You should be able to verify that there will be two completion functions that appears in directories listed on fpath:

1. `/nix/store/$NARHASH-zsh-5.8.1/share/zsh/5.8.1/functions/_curl` which ships with Zsh and has only limited functionality
2. `/run/current-system/sw/share/zsh/site-functions/_curl` which comes from curl.out and is linked into the system environment.

Due to the way the fpath is constructed by the NixOS Zsh module, Zsh will use the limited curl completions that ship with Zsh instead of completions from the package. (And of course this holds for any package that ships completions that extend those that ship with Zsh).

IMHO completions from individual packages should always be preferred and hence should precede the initial Zsh fpath. Do you agree?

**Comment by nrdxp** [comment]
Well when I saw your issue this morning I realized that I didn't have working curl completions either, and all I did was:
```
       curl
  +    curl.out
```
And they worked after a rebuild. I am on a very recent version of nixos-unstable though, so maybe this is only an issue in the stable release?

**Comment by eigengrau** [comment]
Oh boy, the debugging gods really like throwing us the wrench today. But before we doubt our sanity (because the code definitely says that we just use whatever ships with Zsh first), I think I can explain: master has Zsh 5.9, which apparently includes a somewhat more functional, but still limited curl completion as part of Zsh. So I posit that you are still seeing those and not the shiny set of completions that ship with curl. Check this out… 

This is what I expect you will see on your system or any system built from nixpkgs master:

```sh
$ _curl && which _curl
_curl () {
	_gnu_generic || _urls "$@"
}
```

These are the curl completions that ship with Zsh 5.9. They are better than 5.8, but still limited:

```
--data         -d  -- HTTP POST data                                                                                                                                                      
--fail         -f  -- Fail fast with no output on HTTP errors                                                                                                                             
--help         -h  -- Get help for commands                                                                                                                                               
--include      -i  -- Include protocol response headers in the output                                                                                                                     
--output       -o  -- Write to file instead of stdout                                                                                                                                     
--remote-name  -O  -- Write output to a file named as the remote file                                                                                                                     
--silent       -s  -- Silent mode                                                                                                                                                         
--upload-file  -T  -- Transfer local FILE to destination                                                                                                                                  
--user         -u  -- <user-password> Server user and password                                                                                                                            
--user-agent   -A  -- Send User-Agent <name> to server                                                                                                                                    
--verbose      -v  -- Make the operation more talkative                                                                                                                                   
--version      -V  -- Show version number and quit     
```

But if you fix the fpath as per #197503, so that we prefer completions that come from installed packages (and which are found in the system profile), we get the completions that ship with curl:

```
#                                                                                                                                                                                          root@nixos:~/tmp/delme-vm/ > curl -
zsh: do you wish to see all 985 possibilities (247 lines)? y
-
--abstract-unix-socket                     -- Connect via abstract Unix domain socket                                                                                                     
--alt-svc                                  -- Enable alt-svc with this cache file                                                                                                         
--anyauth                                  -- Pick any authentication method                                                                                                              
--append                      -a           -- Append to target file when uploading                                                                                                        
--aws-sigv4                                -- Use AWS V4 signature authentication                                                                                                         
--basic                                    -- Use HTTP Basic Authentication                                                                                                               
--cacert                                   -- CA certificate to verify peer against                                                                                                       
--capath                                   -- CA directory to verify peer against                                                                                                         
--cert                        -E           -- Client certificate file and password                                                                                                        
--cert-status                              -- Verify the status of the server cert via OCSP-staple                                                                                        
--cert-type                                -- Certificate type (DER/PEM/ENG/P12)                                                                                                          
--ciphers                                  -- SSL ciphers to use                                                                                                                          
--compressed                               -- Request compressed response                                                                                                                 
--compressed-ssh                           -- Enable SSH compression                                                                                                                      
--config                      -K           -- Read config from a file                                                                                                                     
--connect-timeout                          -- Maximum time allowed for connection                                                                                                         
--connect-to                               -- Connect to host                                                                                                                             
--continue-at                 -C           -- Resumed transfer offset                                                                                                                     
--cookie                      -b           -- Send cookies from string/file                                                                                                               
--cookie-jar                  -c           -- Write cookies to <filename> after operation                                                                                                 
--create-dirs                              -- Create necessary local directory hierarchy                                                                                                  
--create-file-mode                         -- File mode for created files                                                                                                                 
--crlf                                     -- Convert LF to CRLF in upload                                                                                                                
--crlfile                                  -- Use this CRL list                                                                                                                           
--curves                                   -- (EC) TLS key exchange algorithm(s) to request                                                                                               
--data                        -d           -- HTTP POST data                                                                                                                              
--data-ascii                               -- HTTP POST ASCII data                                                                                                                        
--data-binary                              -- HTTP POST binary data                                                                                                                       
--data-raw                                 -- HTTP POST data, '@' allowed                                                                                                                 
--data-urlencode                           -- HTTP POST data URL encoded                                                                                                                  
--delegation                               -- GSS-API delegation permission                                                                                                               
--digest                                   -- Use HTTP Digest Authentication                                                                                                              
--disable                     -q           -- Disable .curlrc                                                                                                                             
--disable-eprt                             -- Inhibit using EPRT or LPRT                                                                                                                  
--disable-epsv                             -- Inhibit using EPSV                                                                                                                          
--disallow-username-in-url                 -- Disallow username in URL                                                                                                                    
--dns-interface                            -- Interface to use for DNS requests                                                                                                           
--dns-ipv4-addr                            -- IPv4 address to use for DNS requests                                                                                                        
--dns-ipv6-addr                            -- IPv6 address to use for DNS requests                                                                                                        
--dns-servers                              -- DNS server addrs to use                                                                                                                     
--doh-cert-status                          -- Verify the status of the DoH server cert via OCSP-staple                                                                                    
--doh-insecure                             -- Allow insecure DoH server connections                                                                                                       
--doh-url                                  -- Resolve host names over DoH                                                                                                                 
--dump-header                 -D           -- Write the received headers to <filename>                                                                                                    
--egd-file                                 -- EGD socket path for random data                                                                                                             
--engine                                   -- Crypto engine to use                                                                                                                        
--etag-compare                             -- Pass an ETag from a file as a custom header                                                                                                 
--etag-save                                -- Parse ETag from a request and save it to a file                                                                                             
--expect100-timeout                        -- How long to wait for 100-continue                                                                                                           
--fail                        -f           -- Fail fast with no output on HTTP errors                                                                                                     
--fail-early                               -- Fail on first transfer error, do not continue                                                                                               
--fail-with-body                           -- Fail on HTTP errors but save the body                                                                                                       
--false-start                              -- Enable TLS False Start                                                                                                                      
--form-escape                              -- Escape multipart form field/file names using backslash                                                                                      
--form-string                 --form   -F  -- Specify multipart MIME data                                                                                                                 
--ftp-account                              -- Account data string                                                                                                                         
--ftp-alternative-to-user                  -- String to replace USER [name]                                                                                                               
--ftp-create-dirs                          -- Create the remote dirs if not present                                                                                                       
--ftp-method                               -- Control CWD usage                                                                                                                           
--ftp-pasv                                 -- Use PASV/EPSV instead of PORT                                                                                                               
--ftp-port                    -P           -- Use PORT instead of PASV                                                                                                                    
--ftp-pret                                 -- Send PRET before PASV                                                                                                                       
--ftp-skip-pasv-ip                         -- Skip the IP address for PASV                                                                                                                
--ftp-ssl-ccc                              -- Send CCC after authenticating                                                                                                               
--ftp-ssl-ccc-mode                         -- Set CCC mode                                                                                                                                
--ftp-ssl-control                          -- Require SSL/TLS for FTP login, clear for transfer                                                                                           
--get                         -G           -- Put the post data in the URL and use GET                                                                                                    
--globoff                     -g           -- Disable URL sequences and ranges using {} and []                                                                                            
--happy-eyeballs-timeout-ms                -- Time for IPv6 before trying IPv4
… and on and on
```

Does that make sense resp. can you verify that you still only see the curl completions that come with Zsh on your system?

**Comment by nrdxp** [comment]
yeah, looks like I am still using the ones shipped with zsh after all, which is a bit strange considering that they didn't work at all until I added the `curl.out` to my system :shrug: :sweat_smile:
